x-pack/legacy/plugins/epm/server/datasources/create.ts

```diff
@@ -38,30 +38,61 @@ export async function saveDatasourceReferences(options: {
 }) {
   const { savedObjectsClient, pkgkey, toSave } = options;
   const savedObject = await getDatasourceObject({ savedObjectsClient, pkgkey });
-  const savedRefs = savedObject && savedObject.attributes.installed;
-  const mergeRefsReducer = (current: AssetReference[], pending: AssetReference) => {
+  const savedRefs = savedObject?.attributes.package.assets;
+  const mergeRefsReducer = (current: Asset[] = [], pending: Asset) => {
     const hasRef = current.find(c => c.id === pending.id && c.type === pending.type);
     if (!hasRef) current.push(pending);
     return current;
   };
 
-  const toInstall = toSave.reduce(mergeRefsReducer, savedRefs || []);
-
-  await savedObjectsClient.create<DatasourceAttributes>(
-    SAVED_OBJECT_TYPE_DATASOURCES,
-    { installed: toInstall },
-    { id: pkgkey, overwrite: true }
-  );
+  const toInstall = (toSave as Asset[]).reduce(mergeRefsReducer, savedRefs);
+  const datasource: Datasource = createFakeDatasource(pkgkey, toInstall);
+  await savedObjectsClient.create<Datasource>(SAVED_OBJECT_TYPE_DATASOURCES, datasource, {
+    id: pkgkey,
+    overwrite: true,
+  });
 
   return toInstall;
 }
 
 export async function getDatasourceObject(options: {
   savedObjectsClient: SavedObjectsClientContract;
   pkgkey: string;
-}): Promise<DatasourceSavedObject | undefined> {
+}) {
   const { savedObjectsClient, pkgkey } = options;
   return savedObjectsClient
-    .get<DatasourceAttributes>(SAVED_OBJECT_TYPE_DATASOURCES, pkgkey)
+    .get<Datasource>(SAVED_OBJECT_TYPE_DATASOURCES, pkgkey)
     .catch(e => undefined);
 }
+
+function createFakeDatasource(pkgkey: string, assets: Asset[] = []): Datasource {
```

Comment:
Hopefully the name of the function makes the intent clear. Feel free to put better values in here.

Reply:
This will do for now. We should follow up with putting all the values we know in there around the package etc.

An other thing to consider is that when creating a data source, we still don't know man y of the params. For example the pipeline name we should only add when we successfully created the pipeline. Does this mean, these params have to be optional?

Reply:
Yep. We can supply the package info using the EPM saved object which would also ensure the package has been installed before we create a datasource for it. I'm happy to take that ticket/PR as I already wanted to mirror the lessons I learned here about `SavedObjectAttributes` for the EPM state SO.